controller/src/main/java/io/pravega/controller/store/stream/InMemoryStream.java

```diff
@@ -702,6 +704,55 @@ private void handleStreamMetadataExists(final long timestamp, CompletableFuture<
         return result;
     }
 
+    @Override
+    CompletableFuture<Void> createSealedSegmentsRecord(byte[] sealedSegmentRecord) {
+        Preconditions.checkNotNull(sealedSegmentRecord);
+
+        CompletableFuture<Void> result = new CompletableFuture<>();
+
+        synchronized (lock) {
+            this.sealedSegments = new Data<>(sealedSegmentRecord, 0);
+            result.complete(null);
+        }
+        return result;
+    }
+
+    @Override
+    CompletableFuture<Data<Integer>> getSealedSegmentsRecord() {
+        CompletableFuture<Data<Integer>> result = new CompletableFuture<>();
+
+        synchronized (lock) {
+            if (this.sealedSegments == null) {
+                result.completeExceptionally(StoreException.create(StoreException.Type.DATA_NOT_FOUND, getName()));
+            } else {
+                result.complete(copy(sealedSegments));
+            }
+        }
+        return result;
+    }
+
+    @Override
+    CompletableFuture<Void> updateSealedSegmentsRecord(Data<Integer> sealedSegments) {
+        Preconditions.checkNotNull(sealedSegments);
+        Preconditions.checkNotNull(sealedSegments.getData());
+
+        final CompletableFuture<Void> result = new CompletableFuture<>();
+        synchronized (lock) {
+            if (this.sealedSegments == null) {
+                result.completeExceptionally(StoreException.create(StoreException.Type.DATA_NOT_FOUND,
+                        "sealedSegments for stream: " + getName()));
+            } else if (this.sealedSegments.getVersion().equals(sealedSegments.getVersion())) {
+                this.sealedSegments = new Data<>(Arrays.copyOf(sealedSegments.getData(), sealedSegments.getData().length),
```

Comment:
Questions:

1- Could you elaborate on why we need to do this array copy here? 
2- Could you confirm that we have one instance of `SealedSegmentsRecord` per stream?

Reply:
1. we copy the array so that we have an immutable copy of the input so even if the caller changes the array later, our copy does not get affected.
2. yes.